File: 2164-two-best-non-overlapping-events/two-best-non-overlapping-events.py
class Solution:
    def maxTwoEvents(self, events: List[List[int]]) -> int:
        events.sort(key = lambda x : x[0])
        heap = []
        maxSum = 0
        maxCarry = 0
        for event in events:
            while heap and heap[0][0] < event[0]:
                maxCarry = max(maxCarry, heap[0][1])
                heapq.heappop(heap)
            maxSum = max(maxSum, maxCarry + event[2])
            heapq.heappush(heap, (event[1], event[2]))
        return maxSum

File: 2164-two-best-non-overlapping-events/two-best-non-overlapping-events.cpp
class Solution {
public:
    int maxTwoEvents(vector<vector<int>>& events) {
        sort(events.begin(), events.end());
        priority_queue <vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
        int maxSum = 0;
        int maxCarry = 0;
        for (vector<int> event: events){
            while(pq.size() > 0 && pq.top()[0] < event[0]){
                maxCarry = max(maxCarry, pq.top()[1]);
                pq.pop();
            }
            maxSum = max(maxSum, maxCarry + event[2]);
            vector<int> temp;
            temp.push_back(event[1]);
            temp.push_back(event[2]);
            pq.push(temp);
        }
        return maxSum;
    }
};